Docs: Fix Sphinx warnings in sys.rst (#108106)
python/cpython: Doc/library/sys.rst

   +-----------------------------+----------------------------------------------+
   | Attribute                   | Explanation                                  |
   +=============================+==============================================+
   | :attr:`!emscripten_version` | Emscripten version as tuple of ints          |
   |                             | (major, minor, micro), e.g. ``(3, 1, 8)``.   |
   +-----------------------------+----------------------------------------------+
   | :attr:`!runtime`            | Runtime string, e.g. browser user agent,     |
   |                             | ``'Node.js v14.18.2'``, or ``'UNKNOWN'``.    |
   +-----------------------------+----------------------------------------------+
   | :attr:`!pthreads`           | ``True`` if Python is compiled with          |
   |                             | Emscripten pthreads support.                 |
   +-----------------------------+----------------------------------------------+
   | :attr:`!shared_memory`      | ``True`` if Python is compiled with shared   |
   |                             | memory support.                              |

>    +-----------------------------+----------------------------------------------+

Reviewer: I wonder if we could do something like this instead:

```rst
.. data:: _emscripten_info

   A :term:`named tuple` holding information about the environment on the
   *wasm32-emscripten* platform. The named tuple is provisional and may change
   in the future.

   .. attribute:: _emscripten_info.emscripten_version

      Emscripten version as tuple of ints (major, minor, micro), e.g. ``(3, 1, 8)``.

   .. attribute:: _emscripten_info.runtime

      Runtime string, e.g. browser user agent, ``'Node.js v14.18.2'``, or ``'UNKNOWN'``.

   .. attribute:: _emscripten_info.pthreads

      ``True`` if Python is compiled with Emscripten pthreads support.

   .. attribute:: _emscripten_info.shared_memory

      ``True`` if Python is compiled with shared memory support.

   .. availability:: Emscripten.

   .. versionadded:: 3.11
```

I'm not sure if this is how you're "supposed" to use the `.. attribute::` directive (since `_emscripten_info` is module-level data rather than a class). But it seems to render okay for me locally, and there are no warnings while Sphinx is building the docs:

![image](https://github.com/python/cpython/assets/66076021/478a5912-991a-4359-ae06-5083a9fa9838)

Reviewer: (The advantage being that Sphinx adds an anchor for each separate attribute, which could then (I think) be linked to from other pieces of documentation)

Reviewer: Could possibly do the same for `int_info`, `hash_info` and `thread_info` as well (but I'd wait to see if @AA-Turner thinks this general approach is okay before doing so!)

Author: > (The advantage being that Sphinx adds an anchor for each separate attribute, which could then be linked to from other pieces of documentation)

That's a nice side effect. We should do that for the other documented named tuples as well. See also my PR for `float_info`.

Reviewer 2: Seems eminently reasonable; the anchors are formatted correctly, too (``#sys._emscripten_info.emscripten_version``, etc). Also confirmed cross-references would work (e.g. ``:py:attr:`sys._emscripten_info.emscripten_version` ``)

Author: ```suggestion
   .. data:: _emscripten_info

      A :term:`named tuple` holding information about the environment on the
      *wasm32-emscripten* platform.
      The named tuple is provisional and may change in the future.

      .. attribute:: _emscripten_info.emscripten_version

         Emscripten version as tuple of ints (major, minor, micro), e.g. ``(3, 1, 8)``.

      .. attribute:: _emscripten_info.runtime

         Runtime string, e.g. browser user agent, ``'Node.js v14.18.2'``, or ``'UNKNOWN'``.

      .. attribute:: _emscripten_info.pthreads

         ``True`` if Python is compiled with Emscripten pthreads support.

      .. attribute:: _emscripten_info.shared_memory

         ``True`` if Python is compiled with shared memory support.

      .. availability:: Emscripten.

      .. versionadded:: 3.11
```

>    ==============================  ==============================================================================================================
>    attribute                       flag
>    ==============================  ==============================================================================================================

Reviewer 2: For if the *sys.flags* suggestion is accepted
```suggestion
```

>    :attr:`!debug`                  :option:`-d`
>    :attr:`!inspect`                :option:`-i`
>    :attr:`!interactive`            :option:`-i`
>    :attr:`!isolated`               :option:`-I`
>    :attr:`!optimize`               :option:`-O` or :option:`-OO`
>    :attr:`!dont_write_bytecode`    :option:`-B`
>    :attr:`!no_user_site`           :option:`-s`
>    :attr:`!no_site`                :option:`-S`
>    :attr:`!ignore_environment`     :option:`-E`
>    :attr:`!verbose`                :option:`-v`
>    :attr:`!bytes_warning`          :option:`-b`
>    :attr:`!quiet`                  :option:`-q`
>    :attr:`!hash_randomization`     :option:`-R`
>    :attr:`!dev_mode`               :option:`-X dev <-X>` (:ref:`Python Development Mode <devmode>`)
>    :attr:`!utf8_mode`              :option:`-X utf8 <-X>`
>    :attr:`!safe_path`              :option:`-P`
>    :attr:`!int_max_str_digits`     :option:`-X int_max_str_digits <-X>` (:ref:`integer string conversion length limitation <int_max_str_digits>`)
>    :attr:`!warn_default_encoding`  :option:`-X warn_default_encoding <-X>`

Reviewer 2: Following on from Alex's excellent suggestion:

*(I think a later PR perhaps ought fill these in with some more detail, but for now this is just about markup)*

```suggestion
   .. attribute:: flags.debug

      :option:`-d`

   .. attribute:: flags.inspect

      :option:`-i`

   .. attribute:: flags.interactive

      :option:`-i`

   .. attribute:: flags.isolated

      :option:`-I`

   .. attribute:: flags.optimize

      :option:`-O` or :option:`-OO`

   .. attribute:: flags.dont_write_bytecode

      :option:`-B`

   .. attribute:: flags.no_user_site

      :option:`-s`

   .. attribute:: flags.no_site

      :option:`-S`

   .. attribute:: flags.ignore_environment

      :option:`-E`

   .. attribute:: flags.verbose

      :option:`-v`

   .. attribute:: flags.bytes_warning

      :option:`-b`

   .. attribute:: flags.quiet

      :option:`-q`

   .. attribute:: flags.hash_randomization

      :option:`-R`

   .. attribute:: flags.dev_mode

      :option:`-X dev <-X>` (:ref:`Python Development Mode <devmode>`)

   .. attribute:: flags.utf8_mode

      :option:`-X utf8 <-X>`

   .. attribute:: flags.safe_path

      :option:`-P`

   .. attribute:: flags.int_max_str_digits

      :option:`-X int_max_str_digits <-X>` (:ref:`integer string conversion length limitation <int_max_str_digits>`)

   .. attribute:: flags.warn_default_encoding

      :option:`-X warn_default_encoding <-X>`
```

Reviewer: For this one specifically, I think the table format is better. When you're reading the documentation here, you'll just want to follow the links to each option. Having each enumerated as a separate attribute entry feels unnecessarily verbose.

Iff we added a little more description for each entry, I think this would be worth it, but that feels like something that could maybe be postponed to a followup PR

Reviewer 2: What we could do is a list-table similar to #108107, so we're able to retain the canonical links to ``.. attribute:: flags.blah``?

Reviewer: I did not know you could do that with a list-table -- that sounds ideal, yes!

Reviewer 2: List-Table suggestion:
```suggestion
   .. list-table::

      * - .. attribute:: flags.debug
        - :option:`-d`

      * - .. attribute:: flags.inspect
        - :option:`-i`

      * - .. attribute:: flags.interactive
        - :option:`-i`

      * - .. attribute:: flags.isolated
        - :option:`-I`

      * - .. attribute:: flags.optimize
        - :option:`-O` or :option:`-OO`

      * - .. attribute:: flags.dont_write_bytecode
        - :option:`-B`

      * - .. attribute:: flags.no_user_site
        - :option:`-s`

      * - .. attribute:: flags.no_site
        - :option:`-S`

      * - .. attribute:: flags.ignore_environment
        - :option:`-E`

      * - .. attribute:: flags.verbose
        - :option:`-v`

      * - .. attribute:: flags.bytes_warning
        - :option:`-b`

      * - .. attribute:: flags.quiet
        - :option:`-q`

      * - .. attribute:: flags.hash_randomization
        - :option:`-R`

      * - .. attribute:: flags.dev_mode
        - :option:`-X dev <-X>` (:ref:`Python Development Mode <devmode>`)

      * - .. attribute:: flags.utf8_mode
        - :option:`-X utf8 <-X>`

      * - .. attribute:: flags.safe_path
        - :option:`-P`

      * - .. attribute:: flags.int_max_str_digits
        - :option:`-X int_max_str_digits <-X>` (:ref:`integer string conversion length limitation <int_max_str_digits>`)

      * - .. attribute:: flags.warn_default_encoding
        - :option:`-X warn_default_encoding <-X>`
```

Author: I agree with Alex that a table makes sense here. Yay to more `list-table` markups; they are fantastic.

>    ==============================  ==============================================================================================================

Reviewer 2: For if the *sys.flags* suggestion is accepted
```suggestion
```

>    +---------------------+--------------------------------------------------+
>    | attribute           | explanation                                      |
>    +=====================+==================================================+
>    | :attr:`!width`      | width in bits used for hash values               |
>    +---------------------+--------------------------------------------------+
>    | :attr:`!modulus`    | prime modulus P used for numeric hash scheme     |
>    +---------------------+--------------------------------------------------+
>    | :attr:`!inf`        | hash value returned for a positive infinity      |
>    +---------------------+--------------------------------------------------+
>    | :attr:`!nan`        | (this attribute is no longer used)               |
>    +---------------------+--------------------------------------------------+
>    | :attr:`!imag`       | multiplier used for the imaginary part of a      |
>    |                     | complex number                                   |
>    +---------------------+--------------------------------------------------+
>    | :attr:`!algorithm`  | name of the algorithm for hashing of str, bytes, |
>    |                     | and memoryview                                   |
>    +---------------------+--------------------------------------------------+
>    | :attr:`!hash_bits`  | internal output size of the hash algorithm       |
>    +---------------------+--------------------------------------------------+
>    | :attr:`!seed_bits`  | size of the seed key of the hash algorithm       |
>    +---------------------+--------------------------------------------------+

Reviewer 2: ```suggestion
   .. attribute:: hash_info.width

      The width in bits used for hash values

   .. attribute:: hash_info.modulus

      The prime modulus P used for numeric hash scheme

   .. attribute:: hash_info.inf

      The hash value returned for a positive infinity

   .. attribute:: hash_info.nan

      (This attribute is no longer used)

   .. attribute:: hash_info.imag

      The multiplier used for the imaginary part of a complex number

   .. attribute:: hash_info.algorithm

      The name of the algorithm for hashing of str, bytes, and memoryview

   .. attribute:: hash_info.hash_bits

      The internal output size of the hash algorithm

   .. attribute:: hash_info.seed_bits

      The size of the seed key of the hash algorithm
```

>    +----------------------------------------+-----------------------------------------------+
>    | Attribute                              | Explanation                                   |
>    +========================================+===============================================+
>    | :attr:`!bits_per_digit`                | number of bits held in each digit.  Python    |
>    |                                        | integers are stored internally in base        |
>    |                                        | ``2**int_info.bits_per_digit``                |
>    +----------------------------------------+-----------------------------------------------+
>    | :attr:`!sizeof_digit`                  | size in bytes of the C type used to           |
>    |                                        | represent a digit                             |
>    +----------------------------------------+-----------------------------------------------+
>    | :attr:`!default_max_str_digits`        | default value for                             |
>    |                                        | :func:`sys.get_int_max_str_digits` when it    |
>    |                                        | is not otherwise explicitly configured.       |
>    +----------------------------------------+-----------------------------------------------+
>    | :attr:`!str_digits_check_threshold`    | minimum non-zero value for                    |
>    |                                        | :func:`sys.set_int_max_str_digits`,           |
>    |                                        | :envvar:`PYTHONINTMAXSTRDIGITS`, or           |
>    |                                        | :option:`-X int_max_str_digits <-X>`.         |

Reviewer 2: ```suggestion
   .. attribute:: int_info.bits_per_digit

      The number of bits held in each digit.
      Python integers are stored internally in base ``2**int_info.bits_per_digit``.

   .. attribute:: int_info.sizeof_digit

      The size in bytes of the C type used to represent a digit.

   .. attribute:: int_info.default_max_str_digits

      The default value for :func:`sys.get_int_max_str_digits`
      when it is not otherwise explicitly configured.

   .. attribute:: int_info.str_digits_check_threshold

      The minimum non-zero value for :func:`sys.set_int_max_str_digits`,
      :envvar:`PYTHONINTMAXSTRDIGITS`, or :option:`-X int_max_str_digits <-X>`.
```

   A :term:`named tuple` that holds information about Python's internal
   representation of integers.  The attributes are read only.

   .. attribute:: int_info.bits_per_digit

      The number of bits held in each digit.
      Python integers are stored internally in base ``2**int_info.bits_per_digit``.

   .. attribute:: int_info.sizeof_digit

      The size in bytes of the C type used to represent a digit.

   .. attribute:: int_info.default_max_str_digits

      The default value for :func:`sys.get_int_max_str_digits`
      when it is not otherwise explicitly configured.

   .. attribute:: int_info.str_digits_check_threshold

      The minimum non-zero value for :func:`sys.set_int_max_str_digits`,
      :envvar:`PYTHONINTMAXSTRDIGITS`, or :option:`-X int_max_str_digits <-X>`.

   .. versionadded:: 3.1

   .. versionchanged:: 3.11

>       Added :attr:`int_info.default_max_str_digits` and :attr:`int_info.str_digits_check_threshold`.

Reviewer 2: Consider wrapping this? Currently 101 chars.
```suggestion
      Added :attr:`~int_info.default_max_str_digits` and 
      :attr:`~int_info.str_digits_check_threshold`.
```

   A :term:`named tuple` that holds information about Python's internal
   representation of integers.  The attributes are read only.

   .. attribute:: int_info.bits_per_digit

      The number of bits held in each digit.
      Python integers are stored internally in base ``2**int_info.bits_per_digit``.

   .. attribute:: int_info.sizeof_digit

      The size in bytes of the C type used to represent a digit.

   .. attribute:: int_info.default_max_str_digits

      The default value for :func:`sys.get_int_max_str_digits`
      when it is not otherwise explicitly configured.

   .. attribute:: int_info.str_digits_check_threshold

      The minimum non-zero value for :func:`sys.set_int_max_str_digits`,
      :envvar:`PYTHONINTMAXSTRDIGITS`, or :option:`-X int_max_str_digits <-X>`.

   .. versionadded:: 3.1

   .. versionchanged:: 3.11

>       Added :attr:`~int_info.default_max_str_digits` and 

Author: ```suggestion
      Added :attr:`~int_info.default_max_str_digits` and
```

      :file:`Objects/lnotab_notes.txt` for a detailed explanation of how this
      works.
      Per-line events may be disabled for a frame by setting

>       :attr:`!f_trace_lines` to :const:`False` on that frame.

Reviewer: We could possibly link to https://docs.python.org/3/reference/datamodel.html#frame-objects here, which provides some docs on this attribute (and other attributes on types.FrameType)

Author: Hm, yeah, that sounds like an improvement.

Author: But there is already a link to that section a few lines below:

> For more information on code and frame objects, refer to :ref:`types`.

How about adding some more references, like this:

```diff
diff --git a/Doc/library/sys.rst b/Doc/library/sys.rst
index a15bd01a2f..0b4abda060 100644
--- a/Doc/library/sys.rst
+++ b/Doc/library/sys.rst
@@ -1572,7 +1572,7 @@ always available.
       :file:`Objects/lnotab_notes.txt` for a detailed explanation of how this
       works.
       Per-line events may be disabled for a frame by setting
-      :attr:`!f_trace_lines` to :const:`False` on that frame.
+      :attr:`!f_trace_lines` to :const:`False` on that :ref:`frame <frame-objects>`.
 
    ``'return'``
       A function (or other code block) is about to return.  The local trace
@@ -1591,7 +1591,7 @@ always available.
       ``None``; the return value specifies the new local trace function.
       Per-opcode events are not emitted by default: they must be explicitly
       requested by setting :attr:`!f_trace_opcodes` to :const:`True` on the
-      frame.
+      :ref:`frame <frame-objects>`.
 
    Note that as an exception is propagated down the chain of callers, an
    ``'exception'`` event is generated at each level.
```

Reviewer: That LGTM!

   .. versionchanged:: 3.7

      ``'opcode'`` event type added; :attr:`!f_trace_lines` and

>       :attr:`!f_trace_opcodes` attributes added to frames

Reviewer: ```suggestion
      :attr:`!f_trace_opcodes` attributes added to :ref:`frames <frame-objects>`
```

Author: Yeah, I considered this one, but deliberately left it out, as it was only the "version changed" for the `f_trace_opcodes` mentioned earlier, which we already added a frame object link to, so I concluded there was no need for a link here as well :)

>    +------------------+---------------------------------------------------------+
>    | Attribute        | Explanation                                             |
>    +==================+=========================================================+
>    | :attr:`!name`    | Name of the thread implementation:                      |
>    |                  |                                                         |
>    |                  |  * ``'nt'``: Windows threads                            |
>    |                  |  * ``'pthread'``: POSIX threads                         |
>    |                  |  * ``'pthread-stubs'``: stub POSIX threads              |
>    |                  |    (on WebAssembly platforms without threading support) |
>    |                  |  * ``'solaris'``: Solaris threads                       |
>    +------------------+---------------------------------------------------------+
>    | :attr:`!lock`    | Name of the lock implementation:                        |
>    |                  |                                                         |
>    |                  |  * ``'semaphore'``: a lock uses a semaphore             |
>    |                  |  * ``'mutex+cond'``: a lock uses a mutex                |
>    |                  |    and a condition variable                             |
>    |                  |  * ``None`` if this information is unknown              |
>    +------------------+---------------------------------------------------------+
>    | :attr:`!version` | Name and version of the thread library. It is a string, |
>    |                  | or ``None`` if this information is unknown.             |
>    +------------------+---------------------------------------------------------+

Reviewer 2: ```suggestion
   .. attribute:: thread_info.name

      Name of the thread implementation:

      * ``'nt'``: Windows threads
      * ``'pthread'``: POSIX threads
      * ``'pthread-stubs'``: stub POSIX threads
        (on WebAssembly platforms without threading support)
      * ``'solaris'``: Solaris threads

   .. attribute:: thread_info.lock

      Name of the lock implementation:

      * ``'semaphore'``: a lock uses a semaphore
      * ``'mutex+cond'``: a lock uses a mutex and a condition variable
      * ``None`` if this information is unknown

   .. attribute:: thread_info.version

      Name and version of the thread library.
      It is a string, or ``None`` if this information is unknown.
```

      However, if you are writing a library (and do not control in which
      context its code will be executed), be aware that the standard streams
      may be replaced with file-like objects like :class:`io.StringIO` which

>       do not support the :attr:!buffer` attribute.

Reviewer 2: Could I double-check suppressing the cross-reference on this para?

Author: I think we're fine suppressing the link in this case.

   A :term:`named tuple` holding information about the thread
   implementation.

   .. attribute:: thread_info.name

      The name of the thread implementation:

>       * ``"nt"``: Windows threads

Reviewer: This bulleted list might look nicer as a table (same for the one under thread_info.lock)

Author: I prefer the bullet version:

<details>
<summary>Bullets</summary>

<img width="814" alt="Screenshot 2023-08-20 at 08 37 51" src="https://github.com/python/cpython/assets/13780613/17d68044-9b9b-45f8-a3a4-1dd8116fe8e5">

</details>

<details>
<summary>Tables</summary>

<img width="814" alt="Screenshot 2023-08-20 at 08 40 00" src="https://github.com/python/cpython/assets/13780613/d876623b-f861-457e-a387-18c4b27601a7">

</details>

Reviewer: The table version, to me, describes the data more naturally: the data is essentially a one-to-one mapping between the value you could have for `thread_info.name` and the meaning of that particular value. Tables are commonly used for describing mappings in the CPython docs and other docs.

_But_ I don't feel too strongly about it, and the status quo wins in cases of disagreement, so we can leave this if you disagree :)

Reviewer 2: Another alternative would be to use a definition list? (I also prefer bullets, sorry Alex!)

A

Reviewer: > (I also prefer bullets, sorry Alex!)

No worries, I'm obviously outvoted on this one 😄

Reviewer 2: Comparison with a definition list:

<details>
<summary>Bullets</summary>

![image](https://github.com/python/cpython/assets/9087854/d04b4ac9-09be-4dbd-afa8-01c93f942f37)
</details>
<details>
<summary>Tables</summary>

![image](https://github.com/python/cpython/assets/9087854/65488a11-4a45-450b-a115-4eee626a693c)
</details>
<details>
<summary>Definition Lists</summary>

![image](https://github.com/python/cpython/assets/9087854/5ae451a5-5b34-4dce-a811-ae2ec5121801)
</details>

A

Reviewer: I still like the tables best here, but the definition list is also nice in my opinion!

But I'm also happy to just leave it as a bulleted list if others disagree; it's not the biggest issue in the world :)

Author: I have no strong opinion here :) I agree that tables are nice, but I don't like they way they align.

>       * ``"nt"``: Windows threads
>       * ``"pthread"``: POSIX threads
>       * ``"pthread-stubs"``: stub POSIX threads
>         (on WebAssembly platforms without threading support)
>       * ``"solaris"``: Solaris threads

Reviewer 2: ```suggestion
      ``"nt"``
        Windows threads
      ``"pthread"``
        POSIX threads
      ``"pthread-stubs"``
        Stub POSIX threads
        (on WebAssembly platforms without threading support)
      ``"solaris"``
        Solaris threads
```

Reviewer 2: Oh, oops -- forgot to delete these. Sorry for the noise.

   .. attribute:: thread_info.lock

      The name of the lock implementation:

>       * ``"semaphore"``: a lock uses a semaphore
>       * ``"mutex+cond"``: a lock uses a mutex and a condition variable
>       * ``None`` if this information is unknown

Reviewer 2: ```suggestion
      ``"semaphore"``
        A lock uses a semaphore
      ``"mutex+cond"``
        A lock uses a mutex and a condition variable
      ``None``
        When this information is unknown
```
